File: src/main/c/jni-lib/src/events/MessageMapping.cpp
// Copyright (c) 2024 Devexperts LLC.
// SPDX-License-Identifier: MPL-2.0

#include "dxfeed/events/MessageMapping.hpp"
#include "dxfeed/utils/ByteReader.hpp"
#include "dxfeed/utils/ByteWriter.hpp"
#include "dxfeed/utils/JNIUtils.hpp"

namespace dxfeed::jni {
dxfg_message_t *MessageMapping::toMessage(ByteReader &reader) {
    auto *message = new dxfg_message_t();
    message->event_type.clazz = DXFG_EVENT_MESSAGE;
    message->event_symbol = reader.readString();
    message->event_time = reader.readLong();
    message->attachment = bit_cast<void *>(reader.readString());
    return message;
}

void MessageMapping::fromMessage(dxfg_message_t *message, ByteWriter &writer) {
    writer.writeString(message->event_symbol);
    writer.writeInt64_t(message->event_time);
    writer.writeString(bit_cast<const char *>(message->attachment));
}
} // namespace dxfeed::jni

File: src/main/java/com/dxfeed/api/serializers/MessageMapping.java
package com.dxfeed.api.serializers;

import com.dxfeed.api.buffers.ByteBuffer;
import com.dxfeed.api.buffers.ByteReader;
import com.dxfeed.api.utils.XmlMapping;
import com.dxfeed.event.misc.Message;

public class MessageMapping {
  /**
   * https://github.com/dxFeed/dxfeed-graal-native-sdk/blob/main/src/main/c/api/dxfg_events.h#L330
   *
   * typedef struct dxfg_message_t {
   *    dxfg_event_type_t event_type;
   *    const char *event_symbol;
   *    int64_t event_time;
   *    void *attachment;
   * } dxfg_message_t;
   */

  public static void toNative(Message event, ByteBuffer pBytes) {
    pBytes.writeString(event.getEventSymbol()); // 2 + eventSymbolLength
    pBytes.writeLong(event.getEventTime());     // 8
    pBytes.writeString(XmlMapping.toString(event.getAttachment()));
  }

  public static Message fromNative(ByteReader reader) {
    Message message = new Message();
    message.setEventSymbol(reader.readString());
    message.setEventTime(reader.readLong());
    // todo: attachment always is null ?
    Object attachment = message.getAttachment();
    if (attachment != null) {
      message.setAttachment(XmlMapping.fromString(reader.readString(), attachment.getClass()));
    }
    return message;
  }
}
